tracer/src/Datadog.Trace/TraceContext.cs

```diff
@@ -171,47 +161,64 @@ private static void AddSamplingPriorityTags(Span span, int samplingPriority)
             }
         }
 
-        private void PropagateMetadata(ArraySegment<Span> spans)
+        private void PropagateSamplingPriority(Span closedSpan, ArraySegment<Span> spansToWrite, bool containsRootSpan)
         {
-            // Needs to be done for chunks as well, any span can contain the tags.
-            DecorateWithAASMetadata(spans.Array[0]);
-
-            // The agent looks for the sampling priority on the first span that has no parent
-            // Finding those spans is not trivial, so instead we apply the priority to every span
-
-            var samplingPriority = _samplingPriority;
+            if (_samplingPriority == null)
+            {
+                return;
+            }
 
-            if (samplingPriority == null)
+            // This should be the most common case as usually we close the root span last
+            if (closedSpan == RootSpan)
             {
+                AddSamplingPriorityTags(closedSpan, _samplingPriority.Value);
+                _rootSpanSent = true;
                 return;
             }
 
-            // Using a for loop to avoid the boxing allocation on ArraySegment.GetEnumerator
-            for (int i = 0; i < spans.Count; i++)
+            // Normally this use case never happens as the last span closed is the rootspan usually.
+            // So we should have fallen in the previous case.
+            if (containsRootSpan)
             {
-                AddSamplingPriorityTags(spans.Array[i + spans.Offset], samplingPriority.Value);
+                // Using a for loop to avoid the boxing allocation on ArraySegment.GetEnumerator
+                for (var i = 0; i < spansToWrite.Count; i++)
+                {
+                    var span = spansToWrite.Array[i + spansToWrite.Offset];
+                    if (span == RootSpan)
+                    {
+                        AddSamplingPriorityTags(span, _samplingPriority.Value);
+                        _rootSpanSent = true;
+                        return;
+                    }
+                }
+
+                Log.Warning("Root span wasn't found even though expected here.");
+            }
+
+            // Here we must be in the case when rootspan has already been sent or we are in a partial flush
+            // The agent looks for the sampling priority on the first span that has no parent
+            // Finding those spans is not trivial, so instead we apply the priority to every span
```

Comment:
> The agent looks for the sampling priority on the first span that has no parent

^ This old comment was never correct. It was not "the first span that has no parent."

```suggestion
            // Here we must be in the case when rootspan has already been sent or we are in a partial flush.
            // Agent versions < 7.34.0 look for the sampling priority in one of the spans whose parent is not found in the same chunk.
            // If there are multiple orphans, the agent picks one nondeterministically and does not check the others.
            // Finding those spans is not trivial, so instead we apply the priority to every span.
```